codegen/smithy-go-codegen/src/main/java/software/amazon/smithy/go/codegen/integration/Waiters.java

```diff
@@ -0,0 +1,835 @@
+            w.write(
+                    "waiterStateMutator func(context.Context, $P, $P, error) (bool, error)",
+                    inputSymbol, outputSymbol
+            );
+
+            w.addUseImports(SmithyGoDependency.SMITHY_HTTP_TRANSPORT);
+            w.write("requestCloner func(interface{}) interface{}");
+
+            w.write("params $P", inputSymbol);
+            w.write("");
+        });
+    }
+
+    /**
+     * Generates a waiter state mutator function which is used by the waiter retrier Middleware to mutate
+     * waiter state as per the defined logic and returned operation response.
+     *
+     * @param model          the smithy model
+     * @param symbolProvider symbol provider
+     * @param writer         the Gowriter
+     * @param operationShape operation shape on which the waiter is modeled
+     * @param waiterName     the waiter name
+     * @param waiter         the waiter structure that contains info on modeled waiter
+     */
+    private void generateWaiterStateMutator(
+            Model model,
+            SymbolProvider symbolProvider,
+            GoWriter writer,
+            OperationShape operationShape,
+            String waiterName,
+            Waiter waiter
+    ) {
+        StructureShape inputShape = model.expectShape(
+                operationShape.getInput().get(), StructureShape.class
+        );
+        StructureShape outputShape = model.expectShape(
+                operationShape.getOutput().get(), StructureShape.class
+        );
+
+        Symbol inputSymbol = symbolProvider.toSymbol(inputShape);
+        Symbol outputSymbol = symbolProvider.toSymbol(outputShape);
+
+        writer.write("");
+        writer.openBlock("func $L(ctx context.Context, input $P, output $P, err error) (bool, error) {",
+                "}", generateWaiterStateMutatorName(waiterName), inputSymbol, outputSymbol, () -> {
+                    waiter.getAcceptors().forEach(acceptor -> {
+                        writer.write("");
+                        // scope each acceptor to avoid name collisions
+                        Matcher matcher = acceptor.getMatcher();
+                        switch (matcher.getMemberName()) {
+                            case "output":
+                                writer.addUseImports(SmithyGoDependency.GO_JMESPATH);
+                                writer.addUseImports(SmithyGoDependency.FMT);
+
+                                Matcher.OutputMember outputMember = (Matcher.OutputMember) matcher;
+                                String path = outputMember.getValue().getPath();
+                                String expectedValue = outputMember.getValue().getExpected();
+                                PathComparator comparator = outputMember.getValue().getComparator();
+                                writer.openBlock("if err == nil {", "}", () -> {
+                                    writer.write("pathValue, err :=  jmespath.Search($S, output)", path);
```

Comment:
is jmespath just used a place holder until the smithy code gen can be done for jmespath? so SDK doesn't need to rely on this at runtime.

Reply:
If at all possible the SDK should be using code generation for these compariators, with no runtime reflection with jmespath or similar.

Reply:
Created https://github.com/awslabs/smithy-go/issues/238 to track this

Reply:
I was under the impression that out MVP would not require jmespath? Are we saying that we are potentially going to ship with the dependency?

Reply:
I think for now, and come back to it at a later given this is all internal unexported behavior.